flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -537,12 +562,27 @@ else if (!props.forceCheckpoint()) {
 				if (!checkpoint.isDiscarded()) {
 					checkpoint.abortError(new Exception("Failed to trigger checkpoint"));
 				}
+				if(numUnsuccessful > maxUnsuccessfulCheckpoints) {
+					return failExecution(executions);
+				}
 				return new CheckpointTriggerResult(CheckpointDeclineReason.EXCEPTION);
 			}
 
 		} // end trigger lock
 	}
 
+	private CheckpointTriggerResult failExecution(Execution[] executions) {
+		if (currentPeriodicTrigger != null) {
+			currentPeriodicTrigger.cancel();
+			currentPeriodicTrigger = null;
+		}
+		for (Execution execution : executions) {
+			// fail the graph
+			execution.fail(new Throwable("The number of max unsuccessful checkpoints attempts exhausted"));
```

Comment:
I verified the code once again. There is no reference to ExecutionGraph in Checkpointcoordinator and also calling fail on the current Execution actually triggers the restart flow to happen.
Execution#fail()->Marks state to FAILED->vertex#executionFailed()->graph#jobVertexInFinalState(). So you think this way of failing won't work?